Adding MBR partitioning and using it to create a boot image

DrDeano opened the pull request:
Sorry for this being a bit long. I was hoping to split this up into two PR, one for the MBR stuff and one for building a boot image.
But what this adds:
 - Simple MBR partition scheme implementation
 - Refactored makefs to include making a MBR partition table.
 - Refactored some common filesystem operations
 - Removed echfs and parted dependencies (used MBR and FAT32 in house code)
Pls don't look at the build file. It really needs a big refactor, it is a mess, but is working ;) This should be fixed in another PR, contributor friendly :)

SamTebbs33 commented on src/kernel/filesystem/filesystem_common.zig:
```diff
@@ -0,0 +1,173 @@
+const std = @import("std");
+
+// This is the assembly for the FAT bootleader.
```
bootleader -> bootloader

SamTebbs33 commented on src/kernel/filesystem/filesystem_common.zig:
```diff
@@ -0,0 +1,173 @@
+/// Basic boot code that will just print to the scream to insert a bootable image. This is intended
+/// as a place holder to be over written with real boot code if needed.
+/// This assumes the 512 sector size and includes the 0xAA55 boot signature.
+pub const bootsector_boot_code = [512]u8{
```
Wouldn't it be more suitable for this to be in the mbr file as that is what uses it?

DrDeano replied:
It is also used in the FAT32 header, and maybe others for the boot sector.

SamTebbs33 replied:
Hmm, since this is so specialised (we don't yet have other filesystems that use it) and is x86-specific, perhaps it should be moved to somewhere in arch/x86{_64}.

DrDeano replied:
I can do as it is x86 code. I don't know how this boot code is done for ARM, might do some testing for creating a FAT32/MBR partition on a ARM system.

SamTebbs33 replied:
Nice :)

DrDeano replied:
> 
> 
> I can do as it is x86 code. I don't know how this boot code is done for ARM, might do some testing for creating a FAT32/MBR partition on a ARM system.

I'll make this a contributor friendly issue for the ARM boot sector

SamTebbs33 commented on src/kernel/time.zig:
```diff
@@ -0,0 +1,31 @@
+};
+
+pub fn getDateTime() DateTime {
+    return switch (builtin.os.tag) {
```
I'm a little confused as to why this is necessary, when would we not be in freestanding and why do we need to accommodate for other cases?

DrDeano replied:
I use this driver in the build script so runs on the hosts OS, therefore need a non-freestanding version but there isn't one in the standard library yet, see: https://github.com/ziglang/zig/issues/8396

SamTebbs33 replied:
I see, but why use this when building? Is it just because it has a nicer interface than what is currently in std? Using kernel code on the host is a little weird to me.

DrDeano replied:
We could use host stuff like parted and mkfs, but I like using the code we already have that is used in the kernel as we have it so why not use it, plus less dependencies as this can be used on any platform when building as Windows doesn't have parted or mkfs. Plus ensures that what we make using the drivers can be read using the driver (self testing).

DrDeano committed:
Added MBR implementation

This adds the a MBR partition implementation. 
Refactored some file system code to have a common FS helper functions
This also enables the using the MBR and FAT32 to create the boot image removing the echfs dependency.
Moved file system boot sector to common x86 arch.
As this boot sector is written in x86, where would need to be an ARM version, so would be arch dependent.
Added MAP file path
Fixed running the x86_64 port as this wasn't making the boot image, just the elf
